system/mp/test/mptest.cpp

```diff
@@ -521,127 +523,214 @@ void MultiTest(ICommunicator *_comm)
         pexception("Client Exception",e);
     }
 
-    PrintLog("MPTEST: client finished");
+    PrintLog("MPTEST: MultiTest client finished");
 
     server.join();
 
     delete [] targets;
     delete buff;
 }
 
-void MPITest(IGroup *group, ICommunicator *mpicomm)
+void MPRing(IGroup *group, ICommunicator *mpicomm, int iters=0)
 {
-    CMessageBuffer mb;
-    CMessageBuffer mb2;
+    CMessageBuffer smb;
+    CMessageBuffer rmb;
     int myrank = group->rank();
     int numranks = group->ordinality();
 
-    int rnksumtotal = 0;
-    for(int i=0;i<numranks;i++)
-        rnksumtotal += (i+1);
+    if (iters == 0)
+        iters = 1000;
+    if (iters < 10)
+        iters = 10;
 
-    PrintLog("MPTEST: MPITest myrank=%d numranks=%d rnksumtotal=%d", myrank, numranks, rnksumtotal);
+    int pintvl = iters/10;
+    if (pintvl < 5)
+        pintvl = 5;
 
-    // send and recv to/from all others without a send/recv deadlock ...
+    PrintLog("MPTEST: MPRing myrank=%d numranks=%d iters=%d", myrank, numranks, iters);
 
-    mb.clear();
-    mb.append(myrank+1);
+    int *na = new int[numranks];
+    int *pa = new int[numranks];
 
-    rank_t r;
-    int rankval;
-    int ranksum = myrank+1;
+    for (int i=0; i<numranks; i++)
+    {
+        na[i] = (i + 1) % numranks;
+        pa[i] = (i + numranks - 1) % numranks;
+    }
 
-    int left, right;
+    int next = na[myrank];
+    int prev = pa[myrank];
 
-    if (numranks == 2)
+    smb.clear();
+    smb.append(iters);
+    PrintLog("MPTEST: MPRing %d send to rank %d", myrank, next);
+    mpicomm->send(smb,next,MPTAG_TEST);
+
+    while (1)
     {
-        if (myrank == 0)
-        {
-            left = 1;
-            right = 1;
-            PrintLog("MPTEST: MPITest: %d send to rank %d", myrank, right);
-            mpicomm->send(mb,right,MPTAG_TEST);
-
-            mb2.clear();
-            PrintLog("MPTEST: MPITest: %d recv from rank %d", myrank, left);
-            mpicomm->recv(mb2,left,MPTAG_TEST,&r);
-            mb2.read(rankval);
-            ranksum += rankval;
-        }
-        else
+        rank_t r;
+        rmb.clear();
+        if ((iters%pintvl) == 0)
+            PrintLog("MPTEST: MPRing %d recv from rank %d", myrank, prev);
+        mpicomm->recv(rmb,prev,MPTAG_TEST,&r);
+        rmb.read(iters);
+
+        if ((iters%pintvl) == 0)
+            PrintLog("MPTEST: MPRing %d iters = %d", myrank, iters);
+
+        if (iters-- == 0)
+            break;
+
+        // shift source and dest around the ring ...
+
+        for (int i=0; i<numranks; i++)
         {
-            left = 0;
-            right = 0;
-            mb2.clear();
-            PrintLog("MPTEST: MPITest: %d recv from rank %d", myrank, left);
-            mpicomm->recv(mb2,left,MPTAG_TEST,&r);
-            mb2.read(rankval);
-            ranksum += rankval;
-
-            PrintLog("MPTEST: MPITest: %d send to rank %d", myrank, right);
-            mpicomm->send(mb,right,MPTAG_TEST);
+            int n1 = na[i];
```

Comment:
I assume what you are doing here is shifting the next/previous ranks by 1 position at each iteration?

Reply:
Yes - first iteration each rank sends to its next neighbor and receives from its prev neighbor.
Each iteration after that the the next/prev are shifted so that everyone eventually sends/recvs to everyone else.